cli/src/cli.rs

```diff
@@ -1131,24 +1146,65 @@ fn process_deploy(
         bpf_loader::id()
     };
 
-    // Build transactions to calculate fees
-    let mut messages: Vec<&Message> = Vec::new();
-    let (blockhash, fee_calculator, _) = rpc_client
-        .get_recent_blockhash_with_commitment(config.commitment)?
-        .value;
     let minimum_balance = rpc_client.get_minimum_balance_for_rent_exemption(program_data.len())?;
-    let ix = system_instruction::create_account(
-        &config.signers[0].pubkey(),
-        &program_id.pubkey(),
-        minimum_balance.max(1),
-        program_data.len() as u64,
-        &loader_id,
-    );
-    let message = Message::new(&[ix], Some(&config.signers[0].pubkey()));
-    let mut create_account_tx = Transaction::new_unsigned(message);
     let signers = [config.signers[0], program_id];
-    create_account_tx.try_sign(&signers, blockhash)?;
-    messages.push(&create_account_tx.message);
+
+    // Check program account to see if partial initialization has occurred
+    println!("before account check");
+    let initial_instructions = if let Some(account) = rpc_client
+        .get_account_with_commitment(&program_id.pubkey(), config.commitment)?
+        .value
+    {
+        let mut instructions: Vec<Instruction> = vec![];
+        if account.executable {
+            return Err(CliError::DynamicProgramError(
+                "Program account is already executable".to_string(),
+            )
+            .into());
+        }
+        if account.data.is_empty() {
+            instructions.push(system_instruction::allocate(
+                &program_id.pubkey(),
+                program_data.len() as u64,
+            ));
+        }
+        if account.owner != loader_id {
+            instructions.push(system_instruction::assign(&program_id.pubkey(), &loader_id));
+        }
+        if account.lamports < minimum_balance {
+            instructions.push(system_instruction::transfer(
+                &config.signers[0].pubkey(),
+                &program_id.pubkey(),
+                minimum_balance - account.lamports,
+            ));
+        }
+        instructions
+    } else {
+        println!("Should be here");
+        vec![system_instruction::create_account(
+            &config.signers[0].pubkey(),
+            &program_id.pubkey(),
+            minimum_balance.max(1),
```

Comment:
I know this is pre-existing logic, but a one lamport balance is going to be immediately rent-collected

Reply:
True! I don't think we need a `max` check here at all any more.